Orange/widgets/unsupervised/owkmeans.py

```diff
@@ -49,333 +142,277 @@ class Error(widget.OWWidget.Error):
-
-        self.setSizePolicy(QSizePolicy.Preferred, QSizePolicy.Preferred)
-        self.mainArea.setSizePolicy(QSizePolicy.Maximum,
-                                    QSizePolicy.Preferred)
-        self.table_box.setSizePolicy(QSizePolicy.Fixed,
-                                     QSizePolicy.MinimumExpanding)
-        self.table_view.setSizePolicy(QSizePolicy.Preferred,
-                                      QSizePolicy.MinimumExpanding)
-        self.table_box.layout().addWidget(self.table_view)
-        self.hide_show_opt_results()
+        table.selectionModel().selectionChanged.connect(self.select_row)
+        self.mainArea.layout().addWidget(table)
+
+        metrics = QFontMetrics(table.font())
+        table.setColumnWidth(0, metrics.width("9999"))
+        metrics = QFontMetrics(table.horizontalHeader().font())
+        max_width = max(metrics.width(title) for title in self.SCORING_METHODS)
+        for i in range(1, 4):
+            table.setColumnWidth(i, max_width)
 
     def adjustSize(self):
         self.ensurePolished()
         s = self.sizeHint()
         self.resize(s)
 
-    def hide_show_opt_results(self):
-        [self.mainArea.hide, self.mainArea.show][self.optimize_k]()
-        QTimer.singleShot(100, self.adjustSize)
-
-    def sizeHint(self):
-        s = self.controlArea.sizeHint()
-        if self.optimize_k and not self.mainArea.isHidden():
-            s.setWidth(s.width() + self.mainArea.sizeHint().width() +
-                       4 * self.childrenRect().x())
-        return s
-
     def update_k(self):
         self.optimize_k = False
-        self.update()
+        self.apply()
 
     def update_from(self):
         self.k_to = max(self.k_from + 1, self.k_to)
         self.optimize_k = True
-        self.update()
+        self.apply()
 
     def update_to(self):
         self.k_from = min(self.k_from, self.k_to - 1)
         self.optimize_k = True
-        self.update()
-
-    def set_optimization(self):
-        self.updateOptimizationGui()
-        self.update()
+        self.apply()
 
     def check_data_size(self, n, msg_group):
         msg_group.add_message(
             "not_enough_data",
             "Too few ({}) unique data instances for {} clusters")
-        if n > len(self.data):
-            msg_group.not_enough_data(len(self.data), n)
+        data_ok = n <= len(self.data)
+        msg_group.not_enough_data(len(self.data), n, shown=not data_ok)
+        return data_ok
+
+    def _compute_clustering(self, k):
+        # False positives (Setting is not recognized as int)
+        # pylint: disable=invalid-sequence-index
+        try:
+            self.optimization_runs[k] = KMeans(
+                n_clusters=k,
+                init=['random', 'k-means++'][self.smart_init],
+                n_init=self.n_init,
+                max_iter=self.max_iterations,
+                compute_silhouette_score=True)(self.data)
+        except BaseException as exc:
```

Comment:
Is there a special reason why BaseException is needed or would Exception suffice?